common/cpp/src/google_smart_card_common/value_conversion.h

```diff
@@ -27,21 +27,153 @@
 #include <google_smart_card_common/value.h>
 
 namespace google_smart_card {
 
+///////////// Internal helpers ///////////////
+
+namespace internal {
+
+// Visitor of enum type's items that converts a C++ enum value into a string
+// `Value`, by finding the corresponding item among visited ones. This is a
+// non-template class, because enum values are casted to integers.
+class EnumToValueConverter final {
+ public:
+  explicit EnumToValueConverter(int64_t enum_to_convert);
+  EnumToValueConverter(const EnumToValueConverter&) = delete;
+  EnumToValueConverter& operator=(const EnumToValueConverter&) = delete;
+  ~EnumToValueConverter();
+
+  void HandleItem(int64_t enum_item, const char* enum_item_name);
+  bool TakeConvertedValue(const char* type_name, Value* converted_value,
+                          std::string* error_message = nullptr);
+
+ private:
+  const int64_t enum_to_convert_;
+  optional<Value> converted_value_;
+};
+
+// Visitor of enum type's items that converts a string `Value` into a C++ enum
+// value, by finding the corresponding item among visited ones. This is a
+// non-template class, because enum values are casted to integers.
+class EnumFromValueConverter final {
+ public:
+  explicit EnumFromValueConverter(Value value_to_convert);
+  EnumFromValueConverter(const EnumFromValueConverter&) = delete;
+  EnumFromValueConverter& operator=(const EnumFromValueConverter&) = delete;
+  ~EnumFromValueConverter();
+
+  void HandleItem(int64_t enum_item, const char* enum_item_name);
+  bool GetConvertedEnum(const char* type_name, int64_t* converted_enum,
+                        std::string* error_message) const;
+
+ private:
+  const Value value_to_convert_;
+  optional<int64_t> converted_enum_;
+};
+
+}  // namespace internal
+
+///////////// EnumValueDescriptor ////////////
+
+// Class that allows to describe an enum's items and their corresponding string
+// names. Can be used in order to automatically implement conversion of a C++
+// enum to/from a string `Value` object (provided by the global template
+// functions `ConvertToValue()` and `ConvertFromValue()` defined below).
+//
+// Example usage:
+//   enum class Tp { kA, kB };
+//   template <> EnumValueDescriptor<Tp>::Description
+//   EnumValueDescriptor<Tp>::GetDescription() {
+//     return Describe("Tp").WithItem(Tp::kA, "a").WithItem(Tp::kB, "b");
+//   }
+//   ...
+//   ConvertToValue(Tp::kA, ...);
+//   ConvertFromValue(Value("b"), ...);
+template <typename T>
+class EnumValueDescriptor {
+ public:
+  static_assert(std::is_enum<T>::value,
+                "The EnumValueDescriptor parameter must be an enum");
+
+  // Class that should be used for describing items of the enum type. Should be
+  // instantiated via the `Describe()` method.
+  class Description final {
+   public:
+    Description(const char* type_name,
+                internal::EnumToValueConverter* to_value_converter,
+                internal::EnumFromValueConverter* from_value_converter)
```

Comment:
I may be missing something but does Description class needs to know about whether it is EnumToValue or EnumFromValue converter? If it doesn't need to know, then we might have a super class `EnumValueConverter` that both `EnumToValueConverter` and EnumFromValueConverter` inherit from. and we use this class to pass on one instance and verify it is not null.

Reply:
Please, discard.

ooh I missed that this class can be used to convert to and from at the same time.